File: Tarea6/parqueo.m
clc;
clear;

pkg load database;

% Opciones de conexión a la base de datos
dbopts.dbname = 'tarea6';           % Nombre de la base de datos
dbopts.host = 'localhost';           % Dirección del servidor
dbopts.port = '5432';                % Puerto del servidor PostgreSQL
dbopts.user = 'postgres';            % Usuario
dbopts.password = '2008000070*a';             % Contraseña

% Conexión a la base de datos
conn = pq_connect(dbopts);

% Verificación de conexión
if isempty(conn)
    error("Error: No se pudo conectar a la base de datos.");
end

disp("Conexión a la base de datos establecida exitosamente.");

while true
    disp('------ Menú ------');
    disp('1. Ingreso de datos');
    disp('2. Historial de datos');
    disp('3. Borrar datos');
    disp('4. Salir');
    opcion = input('Seleccione una opción (1-4): ');

    switch opcion
        case 1
            % Ingreso de datos
            nombre = input("Ingrese su nombre: ", "s");
            nit = input("Ingrese su NIT: ", "s");
            placa = input("Ingrese su número de placa: ", "s");

            disp("Para los siguientes datos utilice formato de 24 horas (Ej: 16:30).\n");

            % Validación de la hora de entrada
            while true
                entradah = input("Ingrese la hora de entrada (0-23): ");
                if entradah >= 0 && entradah <= 23
                    break;
                else
                    disp("Error: Por favor, ingrese una hora válida.");
                end
            end

            while true
                entradam = input("Ingrese el minuto de entrada (0-59): ");
                if entradam >= 0 && entradam <= 59
                    break;
                else
                    disp("Error: Por favor, ingrese un minuto válido.");
                end
            end

            % Validación de la hora de salida
            while true
                salidah = input("Ingrese la hora de salida (0-23): ");
                if salidah >= entradah && salidah <= 23
                    break;
                else
                    disp("Error: La hora de salida debe ser mayor o igual a la hora de entrada y válida.");
                end
            end

            while true
                salidam = input("Ingrese el minuto de salida (0-59): ");
                if salidam >= 0 && salidam <= 59
                    break;
                else
                    disp("Error: Por favor, ingrese un minuto válido.");
                end
            end

            % Cálculo de tiempo y tarifa
            horas = salidah - entradah;
            if salidam > entradam
                horas = horas + 1;
            end

            if horas < 2
                total = 15;
            else
                total = horas * 20;
            end

            % Mostrar factura
            disp(" ");
            disp("---------------------------------");
            disp("Parqueo del T3");
            disp("FACTURA");
            disp(["Nombre del usuario: ", nombre]);
            disp(["NIT del usuario: ", nit]);
            disp(["Placa del vehículo: ", placa]);
            disp(["Hora de entrada: ", num2str(entradah), ":", num2str(entradam)]);
            disp(["Hora de salida: ", num2str(salidah), ":", num2str(salidam)]);
            disp(["Horas de estadía: ", num2str(horas)]);
            disp(["Monto total a pagar: Q", num2str(total)]);
            disp("---------------------------------");

            % Guardar factura en archivo
            archivo_id = fopen("factura.txt", "a");
            fprintf(archivo_id, "\n---------------------------------\n");
            fprintf(archivo_id, "Parqueo del T3\n");
            fprintf(archivo_id, "FACTURA\n");
            fprintf(archivo_id, "Nombre del usuario: %s\n", nombre);
            fprintf(archivo_id, "NIT del usuario: %s\n", nit);
            fprintf(archivo_id, "Placa del vehículo: %s\n", placa);
            fprintf(archivo_id, "Hora de entrada: %02d:%02d\n", entradah, entradam);
            fprintf(archivo_id, "Hora de salida: %02d:%02d\n", salidah, salidam);
            fprintf(archivo_id, "Horas de estadía: %d\n", horas);
            fprintf(archivo_id, "Monto total a pagar: Q%.2f\n", total);
            fprintf(archivo_id, "---------------------------------\n");
            fclose(archivo_id);

            % Guardar datos en la base de datos
            query = "INSERT INTO parking (nombre, nit, placa, entrada, salida, horas, total) VALUES ($1, $2, $3, $4, $5, $6, $7);";
            params = {nombre, nit, placa, (entradah * 100 + entradam), (salidah * 100 + salidam), horas, total};
            res = pq_exec_params(conn, query, params);

            if isempty(res)
                disp("Error: No se pudo guardar la información en la base de datos.");
            else
                disp("Datos guardados exitosamente en la base de datos.");
            end

        case 2
            % Mostrar historial de datos
            fid = fopen('factura.txt', 'r');
            if fid == -1
                disp('Error: No se pudo abrir el archivo.');
            else
                disp('Contenido del archivo:');
                while ~feof(fid)
                    linea = fgetl(fid);
                    disp(linea);
                end
                fclose(fid);
            end

        case 3
            % Borrar datos
            sql = 'TRUNCATE TABLE parking;';
            pq_exec_params(conn, sql);
            archivo = 'factura.txt';
            if exist(archivo, 'file') == 2
                delete(archivo);
            end
            disp('Datos borrados correctamente.');

        case 4
            % Salir del programa
            disp('Saliendo del programa...');
            break;

        otherwise
            disp('Opción no válida. Intente nuevamente.');
    end
end

% Cerrar conexión a la base de datos
pq_close(conn);

File: Tarea6/Parqueo.py
import psycopg2
from psycopg2 import sql

def input_int(prompt, min_value=None, max_value=None):
    while True:
        try:
            value = int(input(prompt))
            if (min_value is not None and value < min_value) or (max_value is not None and value > max_value):
                print(f"Error: Por favor, ingrese un valor entre {min_value} y {max_value}.")
            else:
                return value
        except ValueError:
            print("Error: Debe ingresar un número válido.")

def menu():
    conn = psycopg2.connect(
        dbname='tarea6',
        host='localhost',
        port='5432',
        user='postgres',
        password='2008000070*a'
    )
    cur = conn.cursor()
    
    while True:
        print('------ Menú ------')
        print('1. Ingreso de datos')
        print('2. Historial de datos')
        print('3. Borrar datos')
        print('4. Salir')
        opcion = input('Seleccione una opción (1-4): ')

        if opcion == '1':
            nombre = input('Ingrese su nombre: ')
            nit = input('Ingrese su número de nit: ')
            placa = input('Ingrese su número de placa: ')
            
            print('Para los siguientes datos utilice formato 24 horas')
            print('Por ejemplo 4:30 pm ingrese primero 16 y luego 30')

            entradah = input_int('Ingrese la hora a la que entró (0-23): ', 0, 23)
            entradam = input_int('Ingrese el minuto al que entró (0-59): ', 0, 59)

            entrada = (entradah * 100) + entradam

            salidah = input_int('Ingrese la hora a la que saldrá (0-23): ', 0, 23)
            while salidah < entradah:
                print('Error: La hora de salida debe ser igual o mayor a la hora de entrada.')
                salidah = input_int('Ingrese la hora a la que saldrá (0-23): ', 0, 23)

            salidam = input_int('Ingrese el minuto al que saldrá (0-59): ', 0, 59)
            while salidah == entradah and salidam <= entradam:
                print('Error: El minuto de salida debe ser mayor que el minuto de entrada si la hora es la misma.')
                salidam = input_int('Ingrese el minuto al que saldrá (0-59): ', 0, 59)

            salida = (salidah * 100) + salidam

            horas = salidah - entradah
            if entradam <= salidam:
                horas += 1

            total = 15 if horas < 2 else horas * 20

            # Mostrar resumen en pantalla
            print('\n---------------------------------')
            print('\nParqueo del T3\n')
            print('FACTURA')
            print(f'\nNombre del usuario: {nombre}')
            print(f'NIT del usuario: {nit}')
            print(f'Placa del vehículo: {placa}')
            print(f'\nHora de entrada: {entradah}:{entradam}')
            print(f'Hora de salida: {salidah}:{salidam}')
            print(f'Horas de estadía: {horas}')
            print(f'\nMonto total a pagar: Q{total:.2f}')
            print('\n---------------------------------')

            # Guardar en archivo factura.txt
            with open('factura.txt', 'a') as archivo:
                archivo.write('\n---------------------------------\n')
                archivo.write('Parqueo del T3\n')
                archivo.write('FACTURA\n')
                archivo.write(f'\nNombre del usuario: {nombre}\n')
                archivo.write(f'NIT del usuario: {nit}\n')
                archivo.write(f'Placa del vehículo: {placa}\n')
                archivo.write(f'\nHora de entrada: {entradah}:{entradam}\n')
                archivo.write(f'Hora de salida: {salidah}:{salidam}\n')
                archivo.write(f'Horas de estadía: {horas}\n')
                archivo.write(f'Monto total a pagar: Q{total:.2f}\n')
                archivo.write('\n---------------------------------\n')

            insert_sql = sql.SQL("""
                INSERT INTO parking (nombre, nit, placa, entrada, salida, horas, total)
                VALUES (%s, %s, %s, %s, %s, %s, %s)
            """)
            cur.execute(insert_sql, (nombre, nit, placa, entrada, salida, horas, total))
            conn.commit()
            print('Datos ingresados correctamente.')

        elif opcion == '2':
            try:
                with open('factura.txt', 'r') as fid:
                    print('Contenido del archivo:')
                    print(fid.read())
            except FileNotFoundError:
                print('El archivo factura.txt no existe.')

        elif opcion == '3':
            truncate_sql = 'TRUNCATE TABLE parking;'
            cur.execute(truncate_sql)
            conn.commit()
            
            try:
                import os
                os.remove('factura.txt')
                print('Datos borrados correctamente.')
            except FileNotFoundError:
                print('El archivo factura.txt no existe.')

        elif opcion == '4':
            print('Saliendo del programa...')
            break

        else:
            print('Opción no válida. Intente nuevamente.')

        input('Presione Enter para continuar...')
    
    cur.close()
    conn.close()

if __name__ == '__main__':
    menu()
